File: examples/gradle/kotlin-js-file-download/kotlin-js-file-download/karakum/src/jsMain/kotlin/js/file/download/karakum/plugins/convertSkippedGenerics.kt
package js.file.download.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isIdentifier
import typescript.isTypeReferenceNode

val convertSkippedGenerics = createPlugin { node, _, render ->
    nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "ArrayBufferView")

        ensure(node.typeArguments == null)

        "${render(node.typeName)}<*>"
    }
}

File: examples/npm-ts/kotlin-js-file-download/karakum/plugins/convertSkippedGenerics.ts
import ts, {type Node} from "typescript"
import type {Context} from "karakum"

export default function (node: Node, context: Context, render: (node: Node) => string) {
    if (
        ts.isTypeReferenceNode(node)

        && ts.isIdentifier(node.typeName)
        && node.typeName.text === "ArrayBufferView"

        && node.typeArguments === undefined
    ) {
        return `${render(node.typeName)}<*>`
    }

    return null
}
